### DISCUSSION TITLE
Error after reading remote json file – Detect whether remote JSON file is compressed

### DISCUSSION THREAD

**Discussion by ericemc3** [duckdb/duckdb#12249] [Ideas]
### What happens?

I can't read a remote Json file, but it is ok if it is read locally.

### To Reproduce

```sql
FROM read_json('https://statistiques-locales.insee.fr/GC_refdata.php?nivgeo=com_courant&extent=fe&lang=fr') ;
```
```text
SQL Error: java.sql.SQLException: Invalid Input Error: Malformed JSON in file "https://statistiques-locales.insee.fr/GC_refdata.php?nivgeo=com_courant&extent=fe&lang=fr", at byte 1 in record/value 2: unexpected character. 
```

If i store first the result from https://statistiques-locales.insee.fr/GC_refdata.php?nivgeo=com_courant&extent=fe&lang=fr -> com.json
`FROM read_json('com.json') ;`
Ok

### OS:

Win11

### DuckDB Version:

10.0.3

### DuckDB Client:

DuckDB cli

### Full Name:

eric mauviere

### Affiliation:

icem7

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a stable release

### Did you include all relevant data sets for reproducing the issue?

Not applicable - the reproduction does not require a data set

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by carlopi** [comment]
The file looks like to be served compressed and with a header of "content-encoding: gzip" (independently of the request).

Here I believe for duckdb to behave correctly in NOT sending "Accept-Header" headers in the request and NOT expecting content to arrive compressed.
We could do better, in the sense that being able to deal also with unreliably servers might be better, but that's more of a feature request, I would probably move this to a GitHub discussion. I think that would be interesting and relatively easy to do, and can possibly save some latency also in general if files can be transfered compressed.

Workarounds are, for now, either fetching and feeding the file to duckdb externally OR set up (possibly via proxy) headers correctly in the remote resource.

**Comment by ericemc3** [comment]
Thank you for that clear explanation. It is logical indeed, as it happens also frequently for random servers to behave this way.

**Comment by szarnyasg** [comment]
Hi @ericemc3 thanks for opening this issue. The problem is indeed what @carlopi is describing.

Two comments:

1) you may pass the `compression = gzip` property to the `read_json` function:

    ```sql
    FROM read_json('https://statistiques-locales.insee.fr/GC_refdata.php?nivgeo=com_courant&extent=fe&lang=fr', compression = gzip);
    ```

    ```text
    ┌──────────────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────────┬────────┐
    │        header        │                                               content                                                │ footer │
    │ struct(contentorde…  │          struct(territories struct(codgeo varchar[], libgeo varchar[], ctrlcodesmd5 uuid))           │ json[] │
    ├──────────────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┤
    │ {'contentOrder': […  │ {'territories': {'codgeo': [01001, 01002, 01004, 01005, 01006, 01007, 01008, 01009, 01010, 01011, …  │ []     │
    └──────────────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────┴────────┘
    ```

2) my reproduction shows one difference: using the `.json` filename format, it also fails with malformed JSON:

    ```sql
    FROM 'out.json';
    ```
    ```
    Invalid Input Error: Malformed JSON in file "out.json", at byte 1 in record/value 2: unexpected character.
    ```

**Comment by wldevries** [comment]
Why was this issue changed into a discussion? I have the same problem when using the online demo app. Adding the argument `compression` does not fix the problem for me. The file loads fine with the local CLI on Windows 10.
```
duckdb> select * from read_json(mileusage.jsonl.gz, compression=gzip, format=newline_delimited);
Invalid Input Error: Malformed JSON in file "mileusage.jsonl.gz", at byte 1 in line 2: unexpected character. 
```

**Reply by carlopi** [reply]
There are two different sides: original bug report is connected to CLI, there it's expected to work, but in this case server is at fault. It has been kept as discussion since I think that might help other find this.

The bug you are reporting, connected to the DuckDB-Wasm online shell is different, and worth of an issue at https://github.com/duckdb/duckdb-wasm. (I think there is already one on this).

**Reply by wldevries** [reply]
https://github.com/duckdb/duckdb-wasm/issues/1172

thanks!
